File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int ct = 0;
        int f[] = new int[26];
        List<Integer> al = new ArrayList<>();
        for(char c : p.toCharArray()){
            int x = c - 'a';
            if(f[x] == 0) ct++;
            f[x]--;
        }
        int l = 0, r = 0;
        while(r < s.length()){
            int in = s.charAt(r) - 'a';
            if(f[in] == 0) ct++;
            f[in]++;
            if(f[in] == 0) ct--;
            if(r - l + 1 < p.length())r++;
            else{
                if(ct == 0) al.add(l);
                int ou = s.charAt(l) - 'a';
                if(f[ou] == 0) ct++;
                f[ou]--;
                if(f[ou] == 0) ct--;
                l++;
                r++;
            }
        }
        return al;
    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int ct = 0;
        vector<int>f(26, 0);
        vector<int>al;
        for(char c : p){
            int x = c - 'a';
            if(f[x] == 0)ct++;
            f[x]--;
        }
        int l = 0, r = 0;
        int n = s.length();
        int m = p.length();
        while(r < n){
            int in = s[r] - 'a';
            if (f[in] == 0) ct++;
            f[in]++;
            if (f[in] == 0) ct--;
            if (r - l + 1 < m)r++;
            else{
                if(ct == 0) al.push_back(l);
                int ou = s[l] - 'a';
                if(f[ou] == 0) ct++;
                f[ou]--;
                if(f[ou] == 0) ct--;
                l++;
                r++;
            }
        }
        return al;
    }
};